src/functional/iterator/embedded.py

```diff
@@ -479,19 +484,23 @@ def execute_shift(
             if p is None:
                 new_entry[i] = index
                 break
-        return pos | {tag: new_entry}
+        # the assertions above confirm pos is incomplete
+        # casting here to avoid duplicating work in a type guard
+        return cast(IncompletePosition, pos) | {tag: new_entry}
 
     assert tag in offset_provider
     offset_implementation = offset_provider[tag]
-    if isinstance(offset_implementation, Dimension):
+    if isinstance(offset_implementation, common.Dimension):
         new_pos = copy.copy(pos)
         if is_int_index(value := new_pos[offset_implementation.value]):
             new_pos[offset_implementation.value] = value + index
         else:
             raise AssertionError()
         return new_pos
     else:
-        assert isinstance(offset_implementation, Connectivity)
+        assert isinstance(
+            offset_implementation, (NeighborTableOffsetProvider, StridedNeighborOffsetProvider)
```

Comment:
This introduces strange coupling, if it's a mypy issue, we need to move it to an `is_connectivity()` function to have this check in only one place.

Reply:
The problem is that `Connectivity` does not have some of the attributes that are accesed later. So either these attributes are at the wrong level or you actually wanted specifically a subclass that has them. I tried not to change the meaning of the code, only reflect the usage in the type annotations.